File: public/js/registration-popup.js
// Registration popup functionality
document.addEventListener('DOMContentLoaded', function() {
    
    // Registration popup elements
    const registrationPopup = document.getElementById('registration-popup');
    const registrationForm = document.getElementById('registration-form');
    const closeRegistrationBtn = document.getElementById('close-registration-popup');
    const submitButton = document.getElementById('registration-submit-btn');
    const emailField = document.getElementById('registration-email');
    const passwordField = document.getElementById('registration-password');
    const confirmPasswordField = document.getElementById('registration-confirm-password');
    const errorMsg = document.getElementById('registration-password-mismatch-error');
    const tokenField = document.getElementById('registration-token');
    const clubSubtitle = document.getElementById('registration-club-subtitle');
    
    // Translation messages
    const messages = {
        nl: {
            passwordsNotMatch: 'Wachtwoorden komen niet overeen',
            networkError: 'Er is een fout opgetreden. Probeer het opnieuw.',
            creating: 'Account aanmaken...',
            allFieldsRequired: 'Alle velden zijn verplicht',
            passwordTooShort: 'Wachtwoord moet minimaal 8 tekens bevatten',
            emailRequired: 'Email is verplicht',
            validEmailRequired: 'Geldig emailadres is verplicht',
            createAccountFor: 'Maak je account aan voor'
        },
        en: {
            passwordsNotMatch: 'Passwords do not match',
            networkError: 'An error occurred. Please try again.',
            creating: 'Creating Account...',
            allFieldsRequired: 'All fields are required',
            passwordTooShort: 'Password must be at least 8 characters long',
            emailRequired: 'Email is required',
            validEmailRequired: 'Valid email address is required',
            createAccountFor: 'Create your account for'
        }
    };
    
    function getMessage(key) {
        const lang = TeambeeUtils.getCurrentLanguage();
        return messages[lang]?.[key] || messages.en[key] || key;
    }
    
    function openRegistrationPopup(token, clubName, prefilledEmail) {
        if (!registrationPopup || !tokenField) return;
        
        // Set the token
        tokenField.value = token;
        
        // Update subtitle with club name
        if (clubSubtitle && clubName) {
            clubSubtitle.textContent = `${getMessage('createAccountFor')} ${clubName}`;
        }
        
        // Prefill email if provided
        if (emailField && prefilledEmail) {
            emailField.value = prefilledEmail;
        }
        
        // Close any other open popups/dropdowns
        TeambeeUtils.closeAllDropdowns();
        
        // Show popup
        TeambeeUtils.disableScroll();
        registrationPopup.classList.remove('hidden');
        
        setTimeout(() => {
            const modalContent = registrationPopup.querySelector('.bg-white');
            if (modalContent) {
                modalContent.style.transform = 'scale(1)';
                modalContent.style.opacity = '1';
            }
        }, 10);
        
        // Focus on email field or password field if email is pre-filled
        setTimeout(() => {
            if (prefilledEmail && passwordField) {
                passwordField.focus();
            } else if (emailField) {
                emailField.focus();
            }
        }, 100);
    }
    
    function closeRegistrationPopup() {
        if (!registrationPopup) return;
        
        const modalContent = registrationPopup.querySelector('.bg-white');
        if (modalContent) {
            modalContent.style.transform = 'scale(0.95)';
            modalContent.style.opacity = '0';
        }
        
        setTimeout(() => {
            registrationPopup.classList.add('hidden');
            TeambeeUtils.enableScroll();
            
            // Clear form
            if (registrationForm) registrationForm.reset();
            if (tokenField) tokenField.value = '';
            
            // Hide messages
            TeambeeUtils.hideMessage(document.getElementById('registration-popup-error'));
            TeambeeUtils.hideMessage(document.getElementById('registration-popup-success'));
            
            // Reset validation
            if (errorMsg) errorMsg.classList.add('hidden');
            if (confirmPasswordField) confirmPasswordField.classList.remove('border-red-500');
            
            // Clear URL parameters
            const newUrl = window.location.pathname + (window.location.pathname.endsWith('/') ? '' : '/');
            window.history.replaceState({}, document.title, newUrl);
        }, 200);
    }
    
    // Check if there's a registration_token in the URL
    const urlParams = new URLSearchParams(window.location.search);
    const registrationToken = urlParams.get('registration_token');
    const prefilledEmail = urlParams.get('email');
    
    if (registrationToken) {
        // Fetch club info for the token
        fetch(`/api/registration/${registrationToken}`, {
            method: 'GET'
        }).then(response => response.json())
        .then(data => {
            if (data.valid) {
                openRegistrationPopup(registrationToken, data.club_name, prefilledEmail);
            } else {
                console.error('Invalid registration token:', data.error);
                window.location.href = '/?error=invalid_token';
            }
        }).catch(error => {
            console.error('Error validating token:', error);
            window.location.href = '/?error=invalid_token';
        });
    }
    
    // Close button handler
    if (closeRegistrationBtn) {
        closeRegistrationBtn.addEventListener('click', function(e) {
            e.preventDefault();
            closeRegistrationPopup();
        });
    }
    
    // Close popup when clicking outside
    if (registrationPopup) {
        registrationPopup.addEventListener('click', function(e) {
            if (e.target === registrationPopup || e.target.classList.contains('backdrop-blur-sm')) {
                closeRegistrationPopup();
            }
        });
    }
    
    // Close popup when pressing Escape key
    document.addEventListener('keydown', function(e) {
        if (e.key === 'Escape' && registrationPopup && !registrationPopup.classList.contains('hidden')) {
            closeRegistrationPopup();
        }
    });
    
    // Form validation
    function validateRegistrationForm() {
        if (!emailField || !passwordField || !confirmPasswordField || !submitButton) return false;
        
        const email = emailField.value.trim();
        const password = passwordField.value;
        const confirmPassword = confirmPasswordField.value;
        let isValid = true;
        
        // Check email
        if (!email) {
            isValid = false;
        } else if (!TeambeeUtils.validateEmail(email)) {
            isValid = false;
        }
        
        // Check password length
        if (password.length < 8) {
            isValid = false;
        }
        
        // Check if passwords match
        if (password && confirmPassword && password !== confirmPassword) {
            confirmPasswordField.setCustomValidity(getMessage('passwordsNotMatch'));
            if (errorMsg) errorMsg.classList.remove('hidden');
            confirmPasswordField.classList.add('border-red-500');
            isValid = false;
        } else {
            confirmPasswordField.setCustomValidity('');
            if (errorMsg) errorMsg.classList.add('hidden');
            confirmPasswordField.classList.remove('border-red-500');
        }
        
        // Check if all fields are filled
        if (!email || !password || !confirmPassword) {
            isValid = false;
        }
        
        // Update submit button state
        if (isValid) {
            submitButton.disabled = false;
            submitButton.classList.remove('disabled:bg-gray-400', 'disabled:cursor-not-allowed');
            submitButton.classList.add('bg-[#3D2E7C]', 'hover:bg-[#3D2E7C]/90');
        } else {
            submitButton.disabled = true;
            submitButton.classList.add('disabled:bg-gray-400', 'disabled:cursor-not-allowed');
            submitButton.classList.remove('hover:bg-[#3D2E7C]/90');
        }
        
        return isValid;
    }
    
    // Add validation listeners
    if (emailField && passwordField && confirmPasswordField) {
        emailField.addEventListener('input', validateRegistrationForm);
        passwordField.addEventListener('input', validateRegistrationForm);
        confirmPasswordField.addEventListener('input', validateRegistrationForm);
        
        // Initial validation
        validateRegistrationForm();
    }
    
    // Form submission handler
    if (registrationForm) {
        registrationForm.addEventListener('submit', async function(e) {
            e.preventDefault();
            
            if (!validateRegistrationForm()) {
                TeambeeUtils.showMessage(document.getElementById('registration-popup-error'), getMessage('allFieldsRequired'));
                return;
            }
            
            const token = tokenField ? tokenField.value : '';
            if (!token) {
                TeambeeUtils.showMessage(document.getElementById('registration-popup-error'), 'Invalid registration token');
                return;
            }
            
            // Show loading state
            submitButton.disabled = true;
            submitButton.textContent = submitButton.dataset.loadingText || getMessage('creating');
            TeambeeUtils.hideMessage(document.getElementById('registration-popup-error'));
            TeambeeUtils.hideMessage(document.getElementById('registration-popup-success'));
            
            try {
                const formData = new FormData(registrationForm);
                formData.append('token', token);
                
                const response = await fetch(`/register/${token}`, {
                    method: 'POST',
                    body: formData,
                    headers: {
                        'X-Requested-With': 'XMLHttpRequest'
                    }
                });
                
                const result = await response.json();
                
                if (result.success) {
                    TeambeeUtils.showMessage(document.getElementById('registration-popup-success'), result.message);
                    
                    // Clear form
                    registrationForm.reset();
                    
                    // Close popup after a short delay and redirect
                    setTimeout(() => {
                        closeRegistrationPopup();
                        
                        if (result.redirect) {
                            window.location.href = result.redirect;
                        }
                    }, 2000);
                } else {
                    TeambeeUtils.showMessage(document.getElementById('registration-popup-error'), result.message);
                    
                    // Clear password fields and focus on first field
                    if (passwordField) passwordField.value = '';
                    if (confirmPasswordField) confirmPasswordField.value = '';
                    if (passwordField) passwordField.focus();
                }
            } catch (error) {
                console.error('Error:', error);
                TeambeeUtils.showMessage(document.getElementById('registration-popup-error'), getMessage('networkError'));
            } finally {
                submitButton.textContent = submitButton.dataset.defaultText || 'Create Account';
                validateRegistrationForm(); // Re-validate to restore button state
            }
        });
    }
}); 

File: app/components/forms/registration_popup.py
"""
Registration Popup Component

Contains the registration popup component for the main page.
"""

from fasthtml.common import *


class RegistrationPopup:
    """Registration popup component."""
    
    def __init__(self, translations=None, versioned_url=None, language="nl"):
        """Initialize the registration popup component."""
        self.translations = translations or {}
        self.versioned_url = versioned_url or (lambda x: x)
        self.language = language
    
    def get_text(self, key: str, default: str = "") -> str:
        """Get translated text for the given key."""
        return self.translations.get(key, default)
    
    def render(self):
        """Render the registration popup."""
        return Div(
            # Modal backdrop
            Div(
                # Modal content
                Div(
                    # Modal header
                    Div(
                        H2(
                            self.get_text("create_account_title", "Create Your Account"),
                            cls="text-2xl font-bold text-[#3D2E7C] mb-2"
                        ),
                        Button(
                            Img(
                                src=self.versioned_url("/static/assets/close.svg"),
                                alt="Close",
                                cls="w-6 h-6 filter brightness-0"
                            ),
                            id="close-registration-popup",
                            cls="text-gray-700 hover:text-gray-900 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:ring-offset-2 rounded-lg"
                        ),
                        cls="flex justify-between items-center mb-6"
                    ),
                    
                    # Modal body
                    Div(
                        P(
                            id="registration-club-subtitle",
                            cls="text-gray-600 mb-6"
                        ),
                        
                        # Error message container
                        Div(
                            id="registration-popup-error",
                            cls="hidden mb-4 p-3 bg-red-50 border border-red-200 rounded-lg text-red-700 text-sm w-full break-words"
                        ),
                        
                        # Success message container
                        Div(
                            id="registration-popup-success",
                            cls="hidden mb-4 p-3 bg-green-50 border border-green-200 rounded-lg text-green-700 text-sm w-full break-words"
                        ),
                        
                        # Registration form
                        Form(
                            # Hidden token field
                            Input(
                                type="hidden",
                                id="registration-token",
                                name="token"
                            ),
                            
                            # Email field
                            Div(
                                Label(
                                    self.get_text("email_label", "Email"),
                                    for_="registration-email",
                                    cls="block text-sm font-medium text-gray-700 mb-1"
                                ),
                                Input(
                                    type="email",
                                    id="registration-email",
                                    name="email",
                                    required=True,
                                    aria_required="true",
                                    autocomplete="email",
                                    cls="w-full px-3 py-2 bg-[#F8F7FB] border border-gray-300 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:border-[#3D2E7C]"
                                ),
                                cls="flex flex-col gap-2 mb-4"
                            ),
                            
                            # Password field
                            Div(
                                Label(
                                    self.get_text("password_label", "Password"),
                                    for_="registration-password",
                                    cls="block text-sm font-medium text-gray-700 mb-1"
                                ),
                                Input(
                                    type="password",
                                    id="registration-password",
                                    name="password",
                                    required=True,
                                    aria_required="true",
                                    autocomplete="new-password",
                                    cls="w-full px-3 py-2 bg-[#F8F7FB] border border-gray-300 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:border-[#3D2E7C]"
                                ),
                                Div(
                                    self.get_text("password_requirement", "Password must be at least 8 characters long"),
                                    cls="text-xs text-gray-500 mt-1"
                                ),
                                cls="flex flex-col gap-2 mb-4"
                            ),
                            
                            # Confirm Password field
                            Div(
                                Label(
                                    self.get_text("confirm_password_label", "Confirm Password"),
                                    for_="registration-confirm-password",
                                    cls="block text-sm font-medium text-gray-700 mb-1"
                                ),
                                Input(
                                    type="password",
                                    id="registration-confirm-password",
                                    name="confirm_password",
                                    required=True,
                                    aria_required="true",
                                    autocomplete="new-password",
                                    cls="w-full px-3 py-2 bg-[#F8F7FB] border border-gray-300 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:border-[#3D2E7C]"
                                ),
                                # Password mismatch error message
                                Div(
                                    self.get_text("passwords_mismatch", "Passwords do not match"),
                                    id="registration-password-mismatch-error",
                                    cls="hidden text-sm text-red-600 mt-1"
                                ),
                                cls="flex flex-col gap-2 mb-6"
                            ),
                            
                            Button(
                                self.get_text("create_account_button", "Create Account"),
                                type="submit",
                                id="registration-submit-btn",
                                data_loading_text=self.get_text("creating_account", "Creating Account..."),
                                data_default_text=self.get_text("create_account_button", "Create Account"),
                                cls="w-full bg-[#3D2E7C] hover:bg-[#3D2E7C]/90 text-white font-medium py-2 px-4 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:ring-offset-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
                            ),
                            
                            id="registration-form",
                            method="post",
                            cls="flex flex-col",
                            role="form"
                        ),
                        cls=""
                    ),
                    
                    cls="bg-white rounded-lg shadow-xl p-6 w-full max-w-md mx-auto transform transition-all duration-300 ease-out scale-95 opacity-0"
                ),
                cls="fixed inset-0 bg-black/50 backdrop-blur-sm flex items-center justify-center p-4 z-50"
            ),
            id="registration-popup",
            cls="hidden"
        ) 